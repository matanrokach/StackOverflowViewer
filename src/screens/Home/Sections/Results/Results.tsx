import React from 'react';

export interface IItem<T> {
  item: T;
  index: number;
}

interface Props<T> {
  data: T[];
  renderItem({item, index}: {item: T, index: number}): React.ReactElement<IItem<T>>;
  keyExtractor(item: T, index: number): string;
}

export const Results = <T extends unknown>({
  data,
  renderItem: Item,
  keyExtractor,
}: Props<T>) => {

  return (
    <>
      {data.map((item, index) => {
        const key = keyExtractor(item, index);

        return (
            <Item {...{item, index}} key={key} />
        );
      })}
    </>
  );
};
