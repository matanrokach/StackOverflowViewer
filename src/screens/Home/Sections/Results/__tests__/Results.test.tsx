import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {Text, TouchableOpacity, View} from 'react-native';
import {Results} from '../Results';

interface IItem {
  text: string;
}

const renderItem = ({item, index}: {item: IItem; index: number}) => (
  <TouchableOpacity>
    <Text>{item.text}</Text>
  </TouchableOpacity>
);

const data = [{text: 'some text 1'}, {text: 'some text 2'}];

const keyExtractor = (item: IItem) => item.text;

describe('Results', () => {
  it('should match snapshot', () => {

		const props = {
			data,
			renderItem,
			keyExtractor,
		};

    const {toJSON} = render(<Results<IItem> {...props} />);

		expect(toJSON()).toMatchSnapshot();

  });

	it('should have N items', () => {

		const props = {
			data,
			renderItem,
			keyExtractor,
		};

		const { getByText } = render(<Results {...props} />);

		props.data.forEach((item => {
			expect(getByText(item.text)).toBeDefined();
		}));

	});

	it('should have clickable items', () => {

		const mockOnPress = jest.fn();

		const renderItem = ({item, index}: {item: IItem; index: number}) => (
			<TouchableOpacity onPress={() => mockOnPress(item)}>
				<Text>{item.text}</Text>
			</TouchableOpacity>
		);

		const props = {
			data,
			renderItem,
			keyExtractor,
		};

		const {getByText} = render(<Results<IItem> {...props} />);

		const listItem = getByText(props.data[0].text);

		fireEvent.press(listItem);

		expect(mockOnPress).toHaveBeenCalledWith(props.data[0]);

	});
	
});
