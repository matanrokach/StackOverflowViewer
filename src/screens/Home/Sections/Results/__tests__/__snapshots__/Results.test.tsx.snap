// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results should match snapshot 1`] = `
Array [
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      some text 1
    </Text>
  </View>,
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      some text 2
    </Text>
  </View>,
]
`;
